File: bindings/php/src/Proven/SafeNetwork.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeNetwork - FFI wrapper for proven_network_* functions.
// All computation delegated to libproven; no logic reimplemented here.

declare(strict_types=1);

namespace Proven;

/**
 * Safe network/IP address operations via libproven FFI.
 *
 * Every method calls the corresponding libproven FFI function.
 */
final class SafeNetwork
{
    /**
     * Parse an IPv4 address string.
     * Returns an array of 4 octets on success, or null on error.
     *
     * @return int[]|null Array of 4 octets, or null on invalid input.
     */
    public static function parseIpv4(string $ip): ?array
    {
        $result = FFI::getLib()->proven_network_parse_ipv4($ip, strlen($ip));
        if ($result->status !== 0) {
            return null;
        }
        $octets = [];
        for ($i = 0; $i < 4; $i++) {
            $octets[] = $result->octets[$i];
        }
        return $octets;
    }

    /**
     * Check if an IPv4 address is in a private range.
     * Requires an IPv4Address struct (use parseIpv4 first and build the struct).
     */
    public static function ipv4IsPrivate(array $octets): bool
    {
        $addr = FFI::getLib()->new('IPv4Address');
        for ($i = 0; $i < 4; $i++) {
            $addr->octets[$i] = $octets[$i];
        }
        return FFI::getLib()->proven_network_ipv4_is_private($addr);
    }

    /**
     * Check if an IPv4 address is a loopback address.
     */
    public static function ipv4IsLoopback(array $octets): bool
    {
        $addr = FFI::getLib()->new('IPv4Address');
        for ($i = 0; $i < 4; $i++) {
            $addr->octets[$i] = $octets[$i];
        }
        return FFI::getLib()->proven_network_ipv4_is_loopback($addr);
    }
}

File: bindings/ruby/lib/proven/safe_network.rb
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
# frozen_string_literal: true

# Safe network operations via FFI to libproven.
# ALL computation delegates to Idris 2 compiled code.

module Proven
  module SafeNetwork
    # Represents a parsed IPv4 address (4 octets).
    IPv4 = Struct.new(:a, :b, :c, :d, keyword_init: true) do
      def to_s
        "#{a}.#{b}.#{c}.#{d}"
      end
    end

    class << self
      # Parse an IPv4 address string.
      # Returns nil on invalid input.
      #
      # @param address [String]
      # @return [IPv4, nil]
      def parse_ipv4(address)
        return nil if address.nil?
        ptr, len = FFI.str_to_ptr(address)

        # proven_network_parse_ipv4 returns IPv4Result { status, address: { octets: [4]u8 } }
        # Total 8 bytes: i32 status + 4xu8 octets
        buf = Fiddle::Pointer.malloc(8, Fiddle::RUBY_FREE)
        sret_types = [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_SIZE_T]
        sret_args = [buf, ptr, len]

        fn = Fiddle::Function.new(
          FFI.handler["proven_network_parse_ipv4"],
          sret_types,
          Fiddle::TYPE_VOID,
          need_gvl: true
        )
        fn.call(*sret_args)

        packed = buf.to_str(8)
        status = packed[0, 4].unpack1("l!")
        return nil unless status == FFI::STATUS_OK

        octets = packed[4, 4].unpack("C4")
        IPv4.new(a: octets[0], b: octets[1], c: octets[2], d: octets[3])
      rescue Fiddle::DLError, TypeError
        nil
      end

      # Check if a string is a valid IPv4 address.
      #
      # @param address [String]
      # @return [Boolean]
      def valid_ipv4?(address)
        !parse_ipv4(address).nil?
      end

      # Check if an IPv4 address is in a private range (RFC 1918).
      #
      # @param address [String]
      # @return [Boolean, nil]
      def private?(address)
        ip = parse_ipv4(address)
        return nil if ip.nil?

        # Pack octets into the IPv4Address struct (4 bytes)
        ipv4_bytes = [ip.a, ip.b, ip.c, ip.d].pack("C4")
        ipv4_ptr = Fiddle::Pointer.to_ptr(ipv4_bytes)

        # proven_network_ipv4_is_private takes IPv4Address by value (4 bytes = i32)
        packed_addr = (ip.a << 24) | (ip.b << 16) | (ip.c << 8) | ip.d
        FFI.invoke_bool(
          "proven_network_ipv4_is_private",
          [Fiddle::TYPE_INT],
          [packed_addr]
        )
      rescue Fiddle::DLError, TypeError
        nil
      end

      # Check if an IPv4 address is a loopback address (127.0.0.0/8).
      #
      # @param address [String]
      # @return [Boolean, nil]
      def loopback?(address)
        ip = parse_ipv4(address)
        return nil if ip.nil?

        packed_addr = (ip.a << 24) | (ip.b << 16) | (ip.c << 8) | ip.d
        FFI.invoke_bool(
          "proven_network_ipv4_is_loopback",
          [Fiddle::TYPE_INT],
          [packed_addr]
        )
      rescue Fiddle::DLError, TypeError
        nil
      end

      # Check if an IPv4 address is public (not private or loopback).
      #
      # @param address [String]
      # @return [Boolean, nil]
      def public?(address)
        return nil unless valid_ipv4?(address)
        priv = private?(address)
        loop = loopback?(address)
        return nil if priv.nil? || loop.nil?
        !priv && !loop
      end
    end
  end
end
